gsy_e_sdk/setups/Grid_Operator_API_scripts/redis_tou_strategy.py

```diff
@@ -0,0 +1,115 @@
+# flake8: noqa
+"""
+Template file to implement Time of Use grid fees
+strategy through the gsy-e-sdk api client using Redis.
+"""
+import os
+import csv
+from time import sleep
+from pendulum import from_format
+from gsy_framework.constants_limits import DATE_TIME_FORMAT, TIME_FORMAT_SECONDS
+from gsy_e_sdk.redis_aggregator import RedisAggregator
+from gsy_e_sdk.redis_market import RedisMarketClient
+from gsy_e_sdk.utils import log_grid_fees_information
+
+current_dir = os.path.dirname(__file__)
+
+MARKET_NAMES = [
+    "Grid",
+    "Community",
+]
+ORACLE_NAME = "dso"
+SLOT_LENGTH = 15  # leave as is
+
+
+class Oracle(RedisAggregator):
+    """Class to represent the Grid Operator client type."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.is_finished = False
+
+    def on_market_cycle(self, market_info):
+        current_market_fee = {}
+        for area_uuid, area_dict in self.latest_grid_tree_flat.items():
+            if area_dict["area_name"] in MARKET_NAMES:
+                self.add_to_batch_commands.last_market_dso_stats(area_uuid)
+                current_market_fee[area_dict["area_name"]] = area_dict[
+                    "current_market_fee"
+                ]
+        next_market_fee = self.set_new_market_fee(market_info)
+        log_grid_fees_information(MARKET_NAMES, current_market_fee, next_market_fee)
+
+    def set_new_market_fee(self, market_info):
+        """Return the market fees for each market for the next time slot."""
+        next_market_fee = {}
+        market_time = from_format(market_info["market_slot"], DATE_TIME_FORMAT)
+        for area_uuid, area_dict in self.latest_grid_tree_flat.items():
+            if area_dict["area_name"] in MARKET_NAMES:
+                next_market_fee[
+                    area_dict["area_name"]
+                ] = calculate_next_slot_market_fee(market_time, area_dict["area_name"])
+
+                self.add_to_batch_commands.grid_fees(
+                    area_uuid=area_uuid,
+                    fee_cents_kwh=next_market_fee[area_dict["area_name"]],
+                )
+        self.execute_batch_commands()
+        return next_market_fee
+
+    def on_event_or_response(self, message):
+        pass
+
+    def on_finish(self, finish_info):
+        self.is_finished = True
+
+
+def read_fee_strategy():
+    "Return a dictionary containing the Time of Use strategy loaded from the CSV input file."
+    with open(
+            os.path.join(current_dir, "resources/ToU.csv"), newline="", encoding="utf-8"
+    ) as csvfile:
+        csv_rows = csv.reader(csvfile, delimiter=" ", quotechar="|")
+        headers = next(csv_rows)[0].split(";")
```

Comment:
I think you can set the delimiter and save the `split()`
```suggestion
        csv_rows = csv.reader(csvfile, delimiter=";", quotechar="|")
        headers = next(csv_rows)[0]
```

Reply:
I tried and got an error, therefore I would propose to keep it as it is.